### ISSUE TITLE
[MRESOURCES-293] Resources should not be read-only

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-resources-plugin#296]
**[Delany](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=delany)** opened **[MRESOURCES-293](https://issues.apache.org/jira/browse/MRESOURCES-293?redirect=false)** and commented

With Maven 3.9.0 I'm getting lots of these warnings

```
[WARNING] Parameter 'resources' is read-only, must not be used in configuration
```

Clearly it shouldn't be marked read-only

https://maven.apache.org/plugins/maven-resources-plugin/examples/resource-directory.html

---

**Affects:** 3.3.0

**Remote Links:**
- [GitHub Pull Request #54
](https://github.com/apache/maven-resources-plugin/pull/54)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[sjaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sjaranowski), [Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Do you set it in plugin configuration? Or in POM build/resources only?

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Well, the examples you pasted are about build/resources, and if you look at mojo page, this parameter is not even documented,  so you delved into "mojo internals" by directly configuring it. In this case IMHO the warning is correct... you are configuring within POM plugin config section an undocumented mojo parameter.

Why do you never use build/resources?

Btw, which mojo are you talking about?

**Comment by jira-importer** [comment]
**[Delany](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=delany)** commented

> With a sufficient number of users of an API,
> it does not matter what you promise in the contract:
> all observable behaviors of your system
> will be depended on by somebody.
> 
> Hyrum's Law

In the end it was a combination of factors that led to me turn off the default-resources execution and create my own discrete executions.

For one thing, the build/resources element is shared among multiple resource mojo executions so i couldn't use it.

To answer the question why a resource was missing from a JAR I had to consider (as a complete novice to Maven)
* multiple build/resources/resource elements in one pom
* inherited build/resources configuration via help-plugin effective-pom
* multiple and often competing include/exclude filters
* mistake in the glob patterns
* filters are laissez-faire/passthru style - no match does not fail the build, and since the logs never show filename I had to add a guard in the form of the verifications-plugin which checked resources were actually copied
* https://maven.apache.org/plugins/maven-resources-plugin/resources-mojo.html#outputDirectory
* https://maven.apache.org/plugins/maven-resources-plugin/resources-mojo.html#overwrite
* https://maven.apache.org/plugins/maven-resources-plugin/resources-mojo.html#usebuildfilters
* did the resources plugin override a generated resource from jaxb/jaxws/xmlbeans. This can get really tricky especially with both jaxb and jaxws plugins
* before sortpom, moving config around could result in a change in plugin execution order, and before I discovered buildplan-plugin I wouldn't have known
* resources were needed from other modules. They were retrieved using ../ syntax. Eventually I replaced this with maven-remote-resources-plugin

No wonder I got this job, no one else wanted it! How many plugins I mention?!

There are too many ways resource copying can go wrong and too few ways to check that it works. I went as far as providing a https://diffoscope.org/ view of all the unpacked artefacts.

So no thanks build/resources. It looked like a cheat so I threw it out and achieved clarity and consistency in the configuration. Today I know exactly what gets copied when and where - no more magic.

Moving forward though - I'm fine with this mojo warning about read-only. I see the intention - its all obvious in retrospect.

For now

```java
-Dorg.slf4j.simpleLogger.log.org.apache.maven.plugin.internal.ReadOnlyPluginParametersValidator=error
```

For later I could try use copy-resources mojo instead.

Perhaps in addition to **Required** and **Optional** parameters the documentation could mention **Prohibited** parameters?

**Comment by jira-importer** [comment]
**[Peter Hicks](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=poggs)** commented

For the benefit of others, I had this problem on Maven 3.9.  I didn't realise the resources need to go under project/build/resources in pom.xml, and I was specifying the whole resources tag under project/build/plugins/plugin/configuration for the maven-resources-plugin.

The issue was made more complex as the documentation at https://maven.apache.org/plugins/maven-resources-plugin/examples/resource-directory.html isn't explicit that the resources tag must be directly under 'build', and not the plugin configuration.  Partly supporting this is a high-ranking hit on Google at https://www.baeldung.com/maven-resources-plugin in which section 3 shows a configuration tag and includes the resources in the plugin configuration.

In summary, including this in pom.xml causes the warning:

`<plugin>`
`    <groupId>org.apache.maven.plugins</groupId>`
`    <artifactId>maven-resources-plugin</artifactId>`
`    <version>${maven-resources-plugin.version}</version>`
`    <configuration>`
`        <propertiesEncoding>UTF-8</propertiesEncoding>`
`        <resources>`
`            <resource>`
`                <directory>${project.basedir}/common</directory>`
`                <includes>`
`                    <include>banner.txt</include>`
`                </includes>`
`                <filtering>true</filtering>`
`            </resource>`
`        </resources>`
`    </configuration>`
`</plugin>`

The correct way to do it, which might need a documentation update to make it explicit, is to move the 'resources' tag up to under the 'build' tag

`<build>`
`    <resources>`
`        <resource>`
`            <directory>${project.basedir}/common</directory>`
`            <includes>`
`                <include>banner.txt</include>`
`            </includes>`
`            <filtering>true</filtering>`
`        </resource>`
`    </resources>`
`    <plugins>`
`        <plugin>`
`            <groupId>org.apache.maven.plugins</groupId>`
`            <artifactId>maven-resources-plugin</artifactId>`
`            <version>${maven-resources-plugin.version}</version>`
`            <configuration>`
`                <propertiesEncoding>UTF-8</propertiesEncoding>`
`            </configuration>`
`        </plugin>`
`    </plugins>`
`</build>`

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Peter is totally right here, **that is the correct way** to configure resources (and sorry for doco being misleading Peter).

On the other hand, I feel that use of configurable (in plugin configuration section) resource may become handy, as some sort of ad-hoc resource copying... still, I consider Delany answer "plugin configuration. We never use build/resources" as plain wrong.

In other words, resources should be configured as Peter wrote, but I can understand there are (rare) cases where they are needed in some (probably custom) invocation of m-resources-p, but that is NOT how we do it everyday. 

Hence, I created PR and will make them read-write, but with a clause: this decision MAY be altered anytime, even reverted as again, proper way is how Peter wrote in comment above.

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Agreement this is doco issue, no code change needed. The already merged PR rolled back in 1946127aac4acb752fd3018dff8c7751434d6ca7.
